### DISCUSSION TITLE
Running fepx on HPC, and nodal stress values

### DISCUSSION THREAD

**Discussion by xshang93** [neperfepx/FEPX#69] [Q&A] [answered]
Hi @mkasemer !

There are some fepx related questions I hope you could help me with.

1. When I run fepx on cluster, it always fails when resources (cores) are allocated from different nodes:
for example, when running a simulation on 16 cores allocated from 7 nodes, this is what I got: 
```
Info   :   - Mesh parameters:
Info   :     > Node number: 361637
Info   :     > Elt  number: 263789
Info   :   [i] Parsed file `simulation.msh'.
Info   : Initializing simulation...
Info   :   - Initializing parallel gather/scatter routines...
Info   :   - Initializing fields from isotropic viscoplastic solution
Info   :     > ITMETHOD_VP: Iteration 1
Info   :       . Solving NL iteration... ERROR  :       . CG_SOLVER_EBE: Convergence failure.
```
However, if I use 16 cores on one same node, the simulation runs without a problem. I'm not sure what causes this, could you please advise?

2. Similar question to #65 : Is it possible to add nodal stress output for fepx? Or is there any possible way to do it in neper, similar to what has been mentioned here  #19 ? For my application, I need continuous stress/strain plots, which seems to be only possible with nodal values in neper. Increasing the mesh density would not be an option for me, as I've already gotten too many meshes to solve.. 

Thank you!

**Comment by rquey** [comment]
>     2. Similar question to [Nodal Stress Value #65](https://github.com/acmelab-ua/FEPX/discussions/65) : Is it possible to add nodal stress output for fepx? Or is there any possible way to do it in neper, similar to what has been mentioned here  [Displacement output #19](https://github.com/acmelab-ua/FEPX/discussions/19) ? For my application, I need continuous stress/strain plots, which seems to be only possible with nodal values in neper. Increasing the mesh density would not be an option for me, as I've already gotten too many meshes to solve..

This is more a Neper task. Not something we can prioritize on our end for now, but I will open an issue for it.

**Reply by xshang93** [reply]
Thanks. Could you maybe point out a way that I can do it on my own? Maybe through some matlab code, or even better, integrate something to neper?

**Reply by rquey** [reply]
"integrate something to neper" I will (on Monday).

**Reply by xshang93** [reply]
> "integrate something to neper" I will (on Monday).

No way! Thats amazing! Thank you so much for looking into it

**Reply by mkasemer** [reply]
I agree that this is more of a Neper task. In FEPX, we output elemental values at elements, and nodal values at nodes. What you describe would be a post-processing step, to be handled by Neper.

**Comment by mkasemer** [comment, marked as answer]
>
> 1. When I run fepx on cluster, it always fails when resources (cores) are allocated from different nodes:
>    for example, when running a simulation on 16 cores allocated from 7 nodes, this is what I got:
> 

Regarding this problem, this is extremely hard to debug. I assume it has something to do with the configuration of your cluster and how jobs are submitted. Are you able to get single-node jobs to complete successfully? If so, this would be a heavy indication that it is a system-specific issue, and I'm not sure I could be of much help with that issue.

**EDIT**: I missed the second portion of your question, where you state that a single-node job completes successfully. I do think, then, that this is a system-specific issue.

**Reply by xshang93** [reply]
Thank you. The bizarre part is I had (although quite rarely) success with multiple nodes simulations. But I agree that it looks more like a system issue. I might reach out to the system admin to try my luck.

**Reply by mkasemer** [reply]
As a note: why are you running on 7x16 processors? For future consideration: partitioning should always be performed on the mesh to facilitate optimized performance of FEPX. When Neper handles partitioning for multi-nodal architecture, both the number of nodes and the number of processors per node must be powers of 2. It would be much better to use either 4x16 or 8x16.

**Reply by rquey** [reply]
In Neper, requesting a total of 7*16=112 partitions using `-part 112` is fine, while `-part 7:16` would not be.

**Reply by xshang93** [reply]
> As a note: why are you running on 7x16 processors? For future consideration: partitioning should always be performed on the mesh to facilitate optimized performance of FEPX. When Neper handles partitioning for multi-nodal architecture, both the number of nodes and the number of processors per node must be powers of 2. It would be much better to use either 4x16 or 8x16.

Sorry if my description above is vague - By saying 16 cores from 7 nodes I actually meant a total of 16 cpus randomly allocated from 7 nodes. How many cpus are from each node is randomly assigned depending on the availability of the system at the moment of allocation. The reason why I did this was that I was merely testing if this simple sample would converge at all. In practice I will prior to use full nodes for high-throughput simulations. With this said, thanks for pointing out the way to partition - I'll definitely look out for it when the time comes.

**Reply by mkasemer** [reply]
I have never seen job submission in this way... generally you try to reserve all CPUs on a given node before asking for CPUs on another node. Node-to-node communication is the biggest bottleneck in terms of communication between partitions, and should be avoided.

**Reply by xshang93** [reply]
I agree with you. I assume the system behaves this way because it's trying to make use of all possible free resources. I'm quite new to hpc but I can already see problems with this, besides what you have mentioned, - that even the wait time too start the job is longer than if I request all resources to be from one single node..
